label_studio/fsm/state_manager.py

```
@@ -0,0 +1,385 @@
+                # Get denormalized fields from the state model itself
+                denormalized_fields = {}
+                if hasattr(state_model, 'get_denormalized_fields'):
+                    denormalized_fields = state_model.get_denormalized_fields(entity)
+
+                # Get organization from user's active organization
+                organization_id = (
+                    user.active_organization.id
+                    if user and hasattr(user, 'active_organization') and user.active_organization
+                    else None
+                )
+
+                new_state_record = state_model.objects.create(
+                    **{entity._meta.model_name: entity},
+                    state=new_state,
+                    previous_state=current_state,
+                    transition_name=transition_name,
+                    triggered_by=user,
+                    context_data=context or {},
+                    reason=reason,
+                    organization_id=organization_id,
+                    **denormalized_fields,
+                )
+
+                # Update cache with new state
+                cache_key = cls.get_cache_key(entity)
+                cache.set(cache_key, new_state, cls.CACHE_TTL)
+
+                logger.info(
+                    f'State transition successful: {entity._meta.label_lower} {entity.pk} '
+                    f'now in state {new_state} (record ID: {new_state_record.id})'
+                )
+                return True
+
+        except Exception as e:
+            # On failure, invalidate potentially stale cache
+            cache_key = cls.get_cache_key(entity)
+            cache.delete(cache_key)
+            logger.error(
+                f'State transition failed for {entity._meta.label_lower} {entity.pk}: '
+                f'{current_state} → {new_state}: {e}'
+            )
+            raise StateManagerError(f'Failed to transition state: {e}') from e
```

We aren't returning False if it failed, we are raising an error. Might be okay to raise an error and not return anything from this function. We'll need to update `BaseTransition.execute` that expects a boolean.

---

Good catch, I was pondering a concept to make the execution log or raise (reason being that in some circumstances, for example if we use this to backfill states, we wouldn't want it to fail hard, just log and move on). So I think in that regard we may want to tool the behaviour accordingly, and in the case of log only mode it should return whether it succeeded.